### ISSUE TITLE
Trouble accessing the execution data in the directory created during Hyperparameter Tuning

### ISSUE THREAD

**Issue by ChristanBarnard** [keras-team/keras-tuner#526]
I am tuning an ANN model using the Keras Tuner, using the Bayesian optimizer as follows:

```
tuner = BayesianOptimization (
    build_model,
    objective=kt.Objective('val_auc', direction='max'),
    max_trials=30,
    executions_per_trial=10,
    directory='my_dir',
    project_name='silent_sound_ann_bayesian_search_1.5')
```

The code runs, but I want to access the data stored for each of the 10 executions per trial. I can access the checkpoints and trial.json files for each trial via the Colab file structure in ```directory='my_dir'```, but it doesn't seem like the AUC score for each of the 10 executions per trial is logged. 

There are 30 trial.json files in ```directory='my_dir'```, one for each hyperparameter configuration tested. But each of those configurations was executed 10 times, which means there should be 10 AUC values stored in each trial.json file. 

I want to plot box plots of the AUC scores of the executions for each hyperparameter configuration, and doing so requires access to the scores of each execution. Any suggestions on where I might find that data? It doesn't seem to be stored in the oracle.json or the tuner.json files either. Or a more efficient way of capturing the execution scores as the model runs each trial?

**Comment by haifeng-jin** [comment]
This might be a bug.
Would you put a colab notebook for reproducing the bug?

**Comment by roads** [comment]
Perhaps things have moved on and this is the intended behavior, but I'm encountering a similar issue.

## Colab Notebook
Here is a colab notebook containing a minimal working example that demonstrates the issue:
https://colab.research.google.com/drive/1wUXEJP3C7KxtDUtasCWjUM7u0tI-GSSV?usp=sharing

## Tracing the Issue
If this is not the intended behavior, I believe the issue can be traced to the following sequence of events inside `BaseTuner.search()`:

1. Metric results for all executions are retrieved via `run_trial` and stored in a list called `results`.
    * `results = self.run_trial(trial, *fit_args, **fit_kwargs)`
    * line 179 in `BaseTuner`
    * At this point, the data associated with each execution is available and has not been averaged.
2. `results` is consumed by the function `convert_to_metrics_dict`, which calls `average_metrics_dicts` and returns a new dictionary comprised of execution averages.
    * starting on line 192 in `BaseTuner`
3. The newly created "averaged metrics" dictionary is provided as an argument to `self.oracle.update_trial`, which sets off a chain of events that stores the metrics in a `MetricHistory` object attached to the `Tuner` object (e.g., stored at `tuner.oracle.trials['0'].metrics.metrics['loss']._observations`).
    * Since individual execution data is not given to `update_trial`, the non-averaged execution data is effectively lost to the `Tuner` object at this point and is not stored in the `trial.json` files.

## Possible Solution
Could it be as simple as modifying `BaseTuner.search()` so that `update_trial` is called inside a loop when `results` is a list? 

Something like:

```python
# Insert after `else:` on line 191 in base_tuner.py.
if isinstance(results, list):
    for result in results:
        self.oracle.update_trial(
            trial.trial_id,
            # Convert to dictionary before calling `update_trial()`
            # to pass it from gRPC.
            tuner_utils.convert_to_metrics_dict(
                result, self.oracle.objective, "Tuner.run_trial()"
            ),
        )
else:
    # Normal call without for loop.
```

If I make this modification to a local copy of keras-tuner:
* Information for each execution is stored inside the `MetricHistory` object that is part of the tuner.
    * The `Oracle` and `MetricHistory` classes have the functionality for correctly appending data from multiple executions.
* The `trial.json` files now contain the individual execution data.
* The average-over-executions is still reported to the oracle so a trial gets the same score as before.
    * When `self.on_trial_end(trial)` is called near the end of `search`, this sets off a function dive that ends with `MetricHistory.get_best_value()` which computes a mean over all executions.
        * line 101: `values = list(obs.mean() for obs in self._observations.values())`
* All the keras-tuner tests inside `keras_tuner/engine` pass except `tuner_test.py::test_tunable_false_hypermodel`, which generates an OOM error since my GPU isn't big enough to handle something inside the test.
